cli/pcluster/config/s3_param_types.py

```diff
@@ -0,0 +1,241 @@
+        for param_key, param_definition in self.definition.get("params").items():
+            param_type = param_definition.get("type", Param)
+            param = param_type(
+                self.key, self.label, param_key, param_definition, self.pcluster_config, owner_section=self
+            ).from_storage(storage_params)
+            self.add_param(param)
+        return self
+
+    def to_storage(self, storage_params):
+        """Write the section into storage params."""
+        for param_key, _ in self.definition.get("params").items():
+            param = self.get_param(param_key)
+            if param:
+                param.to_storage(storage_params)
+
+    def has_metadata(self):
+        """No metadata must be stored in CloudFormation for S3 Sections."""
+        return False
+
+
+class S3Param(Param):
+    """Base class to manage configuration parameters stored into an S3 bucket."""
+
+    def from_file(self, config_parser):
+        """Load the param value from configuration file."""
+        super(S3Param, self).from_file(config_parser)
+        self.value = self._parse_value(self.value)
+        return self
+
+    def from_storage(self, storage_params):
+        """Load the param from the provided S3 storage params dict."""
+        storage_value = _get_storage_subdict(self, storage_params.s3_params).get(
+            self.get_storage_key(), self.get_default_value()
+        )
+        self.value = storage_value
+        return self
+
+    def to_storage(self, storage_params):
+        """Store the param into the provided S3 storage params dict."""
+        _get_storage_subdict(self, storage_params.s3_params)[self.get_storage_key()] = self.value
+
+    def _parse_value(self, value):
+        """Parse the value from config file, converting to the needed type for the specific param."""
+        # No conversion is applied at base level
+        return value
+
+
+class IntS3Param(S3Param):
+    """Base S3Param to manage int parameters."""
+
+    def _parse_value(self, value):
+        """Parse the value from config file, converting to the needed type for the specific param."""
+        # Convert value to Int
+        return int(value)
+
+
+class BooleanS3Param(S3Param):
+    """Base S3Param to manage boolean parameters."""
+
+    def _parse_value(self, value):
+        """Parse the value from config file, converting to the needed type for the specific param."""
+        # Convert value to boolean
+        return value and value.lower().strip() == "true"
+
+
+class FloatS3Param(S3Param):
+    """Base S3Param to manage float parameters."""
+
+    def _parse_value(self, value):
+        """Parse the value from config file, converting to the needed type for the specific param."""
+        # Convert value to Float
+        return float(value) if value else None
+
+
+class ComputeInstanceTypeS3Param(S3Param):
+    """S3Param to manage compute instance type in compute_resource sections."""
+
+    def refresh(self):
```

Comment:
It's not clear to me why are we initializing internal parameters in the "refresh" method instead of the _from_file one?

Or I mean, we should do it in both the cases.

Reply:
Instead of using the ComputeInstanceTypeS3Param to initialize all the other parameters I'd define a custom VcpuParam, GpuParam, EnableEfaParam etc, able to detect from where they have to take their values.

Every param type should know how to initialize itself (according to the instance type for example, and we can store the instance_type info somewhere) but I'd avoid to put all the initialization logic of all the parameters inside this one. 

With the current code we're hiding this detail. We're saying it is automatically populated but it's not clear how. Instead by using a "GpuParam" etc we're expliciting it.

Let's talk about it if my comments are not clear.

Reply:
> It's not clear to me why are we initializing internal parameters in the "refresh" method instead of the _from_file one?

The ``refresh`` method is a convenient place to recalculate a parameter, since it's called every time the configuration is loaded or updated, from file or from storage.

> Instead of using the ComputeInstanceTypeS3Param to initialize all the other parameters I'd define a custom VcpuParam, GpuParam, EnableEfaParam etc, able to detect from where they have to take their values.

The logic to update ``vcpus``, ``gpus``, ``enable_efa`` etc. is all dependent on the value of the ``compute_instance_type`` param. Defining the refreshing logic inside specific subclasses would mean having more code to maintain, and decentralizing the logic, so that, for instance, multiple calls to ``describe_instance_types`` would be needed. To avoid the multiple calls we could decide to cache the information somewhere, hence complicating the logic and introducing a precedence order in the ``refresh`` calls.